src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,93 @@
 package mate.academy.lib;
 
+import java.io.File;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
 public class Injector {
+    private static final String IMPL_PACKAGE = "mate.academy.service.impl";
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (instances.containsKey(interfaceClazz)) {
+            return instances.get(interfaceClazz);
+        }
+        return createInstance(interfaceClazz);
+    }
+
+    private Object createInstance(Class<?> clazz) {
+        try {
+            if (clazz.isInterface()) {
+                clazz = findImplementation(clazz);
+            }
+
+            if (!clazz.isAnnotationPresent(Component.class)) {
+                throw new RuntimeException("Class " + clazz.getName()
+                        + " is missing @Component annotation");
+            }
+
+            Constructor<?> constructor = clazz.getDeclaredConstructor();
+            Object instance = constructor.newInstance();
+            injectDependencies(instance);
+            instances.put(clazz, instance);
+            return instance;
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create an instance of class "
+                    + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        try {
+            String path = IMPL_PACKAGE.replace('.', '/');
+            URL resource = ClassLoader.getSystemClassLoader().getResource(path);
+            if (resource == null) {
+                throw new RuntimeException("Package " + IMPL_PACKAGE + " not found");
+            }
+
+            File directory = new File(resource.toURI());
+            if (!directory.exists()) {
+                throw new RuntimeException("Directory " + directory.getAbsolutePath()
+                        + " not found");
+            }
+
+            for (File file : Objects.requireNonNull(directory.listFiles())) {
+                if (file.getName().endsWith(".class")) {
+                    String className = IMPL_PACKAGE + "." + file.getName()
+                            .replace(".class", "");
+                    Class<?> implClass = Class.forName(className);
+
+                    if (interfaceClazz.isAssignableFrom(implClass)
+                            && implClass.isAnnotationPresent(Component.class)) {
```

Comment:
The implementation search checks if the class is assignable from the interface and if it has the `@Component` annotation. Ensure that the implementation classes are correctly annotated and located in the specified package.